testing/src/test/java/io/stargate/it/cql/TokenAwareTest.java

```diff
@@ -0,0 +1,47 @@
+package io.stargate.it.cql;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.datastax.oss.driver.api.core.CqlIdentifier;
+import com.datastax.oss.driver.api.core.CqlSession;
+import com.datastax.oss.driver.api.core.ProtocolVersion;
+import com.datastax.oss.driver.api.core.metadata.Node;
+import com.datastax.oss.driver.api.core.metadata.TokenMap;
+import com.datastax.oss.driver.api.core.type.codec.TypeCodecs;
+import io.stargate.it.BaseOsgiIntegrationTest;
+import io.stargate.it.driver.CqlSessionExtension;
+import io.stargate.it.driver.TestKeyspace;
+import io.stargate.it.storage.StargateEnvironmentInfo;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.UUID;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(CqlSessionExtension.class)
+public class TokenAwareTest extends BaseOsgiIntegrationTest {
+  @Test
+  @DisplayName("Should use all Stargate addresses when using token-aware load balancing")
+  public void testTokenMapDistribution(
+      CqlSession session, StargateEnvironmentInfo stargate, @TestKeyspace CqlIdentifier keyspace) {
+    assertThat(session.getMetadata().getTokenMap()).isPresent();
+    final TokenMap tokenMap = session.getMetadata().getTokenMap().get();
+
+    Collection<Node> expectedReplicasTried = session.getMetadata().getNodes().values();
+    assertThat(stargate.nodes()).hasSize(expectedReplicasTried.size());
+
+    Set<Node> allReplicasTried = new HashSet<>();
+    for (int i = 0; i < 2 * expectedReplicasTried.size(); ++i) {
+      Set<Node> replicas =
+          tokenMap.getReplicas(
+              keyspace, TypeCodecs.UUID.encode(UUID.randomUUID(), ProtocolVersion.DEFAULT));
+      assertThat(replicas).isNotEmpty();
+      Node replica = replicas.iterator().next();
+      allReplicasTried.add(replica);
+    }
```

Comment:
I wonder if increasing the number of tokens is going to help with making this distribution more uniform. Can we scale the number of tokens arbitrarily? Do they affect anything in Stargate?

I'm curious what the this distribution is like... I assume there may be some skewness in it since tokens are derived from IP addresses. A small number of tokens will probably allocate larger ring segments to some nodes than to others... Am I mixed up here? :)

Reply:
Basically, what I'm trying to say is if we run the loop on like 36 longer and longer and count the number of hits per node, are these counts going to converge or diverge?

Reply:
The distribution converged to within <25% of each other with 8 random tokens. It's never going to be perfect and we might have random seed collisions, but I think you're right we should probably increase the default and maybe make it configurable.

```java
c.num_tokens = Integer.getInteger("stargate.num_tokens", 32);
```

I had test for doing testing that. Let me run it with some different numbers and see if I can make the distribution a bit more uniform with a larger default.

```java
  @Test
  public void queryDistribution(CqlSession session) {
    final int numQueries = 100;
    PreparedStatement preparedStatement = session.prepare("INSERT INTO test (k, v) VALUES (?, ?)");

    class Counter {
      public int count = 0;

      void increment() {
        ++count;
      }
    }

    Map<Node, Counter> counts = new HashMap<>();
    for (int i = 0; i < numQueries; ++i) {
      ResultSet rs = session.execute(preparedStatement.bind(UUID.randomUUID(), i));
      counts
          .computeIfAbsent(rs.getExecutionInfo().getCoordinator(), n -> new Counter())
          .increment();
    }

    final int expectedCount = numQueries / counts.size();
    for (Map.Entry<Node, Counter> entry : counts.entrySet()) {
      assertThat(entry.getValue().count).isCloseTo(expectedCount, Percentage.withPercentage(25));
    }
  }
```

Reply:
> The distribution converged to within <25% of each other with 8 random tokens.

25% variance might be significant in high load situations... some nodes might be doing ok, while others hitting capacity :thinking: 

+1 to making the number of tokens configurable and re-testing a few different settings, please... perhaps 8, 16, 32?

Reply:
From my POV the dangerous thing is that we cannot control the distribution intelligently. Even though it is deterministic, if one particular set of IP addresses results in an imbalanced cluster, we do not appear to have a mechanism to rebalance it... would adding new nodes with the same token count change the picture? Is the load distribution smooth over the "add node" operation or could it result in imbalances with adding one extra node?

Sorry, many question, not suggestions :) I'm open to discussing this in real time :)

Reply:
For the CDC use case, we cannot achieve the same ordering of events in Stargate and Kafka - they will have different partitions.
The same PK can end up being processing by one stargate node or another. Thus, the events for the same partition key are sent from different Kafka producers resulting in out-of-order events.
Clients can order it if they want to use, for example using moving windows and write-timestamps.
It would be great if clients that the primary or only use case is using CDC could enforce that partition alignment between stargate and Kafka. I wonder if implementing it would be feasible. wdyt?

Reply:
1:1 alignment with Kafka would prevent Stargate from scaling independently, wouldn't it?

Reply:
A possible alternative to token allocation: Choose the starting point randomly (deterministically) and them choose the next N tokens with equal ranges (as opposed to choosing each token randomly).

Reply:
Here's results of running different number of tokens.

```
### 8 tokens ###

Run #1
Counts: (8 tokens)
Node(endPoint=/127.0.0.11:9043, hostId=71e7ab19-6c3e-48c2-8669-a60b045bf1c9, hashCode=69a1ea90) handled 316893 queries of 1000000 which is 31.689300% of the total query load (or 4.932005% from expected)
Node(endPoint=/127.0.0.12:9043, hostId=abf01fa6-52a3-4b3f-a16b-7aef60650d14, hashCode=6d141e6e) handled 312188 queries of 1000000 which is 31.218800% of the total query load (or 6.343506% from expected)
Node(endPoint=/127.0.0.13:9043, hostId=abf756d2-b805-48ed-8d80-2fed049b5e25, hashCode=7a32274a) handled 370919 queries of 1000000 which is 37.091900% of the total query load (or -11.275811% from expected)

Run #2
Counts: (8 tokens)
Node(endPoint=/127.0.0.11:9043, hostId=b8726419-c40b-42e6-bec5-90cb274aafdb, hashCode=4d4d7877) handled 316626 queries of 1000000 which is 31.662600% of the total query load (or 5.012105% from expected)
Node(endPoint=/127.0.0.12:9043, hostId=eafbe3a3-0068-485c-9f1d-df5829515a67, hashCode=37faf3c7) handled 312680 queries of 1000000 which is 31.268000% of the total query load (or 6.195906% from expected)
Node(endPoint=/127.0.0.13:9043, hostId=4839081c-1d56-42c1-9dad-ab08fbe8cb5a, hashCode=5b77189) handled 370694 queries of 1000000 which is 37.069400% of the total query load (or -11.208311% from expected

Run #3
Counts: (8 tokens)
Node(endPoint=/127.0.0.11:9043, hostId=7d7ea236-f097-4925-a394-cd84a89e5f2c, hashCode=55b6fc69) handled 316346 queries of 1000000 which is 31.634600% of the total query load (or 5.096105% from expected)
Node(endPoint=/127.0.0.12:9043, hostId=bd7bf09e-e108-474f-9797-9bdb98332797, hashCode=5c8e3c61) handled 312241 queries of 1000000 which is 31.224100% of the total query load (or 6.327606% from expected)
Node(endPoint=/127.0.0.13:9043, hostId=c8580fd2-315e-4702-9450-46f7136c1ca6, hashCode=59acd670) handled 371413 queries of 1000000 which is 37.141300% of the total query load (or -11.424011% from expected)

### 16 tokens ###

Run #1
Counts: (16 tokens)
Node(endPoint=/127.0.0.11:9043, hostId=f0255a87-e658-46a4-975a-950f286be847, hashCode=3dd784c4) handled 284279 queries of 1000000 which is 28.427900% of the total query load (or 14.716215% from expected)
Node(endPoint=/127.0.0.12:9043, hostId=e51ff24b-5408-4f4c-910b-4231db005fad, hashCode=61eddf68) handled 386282 queries of 1000000 which is 38.628200% of the total query load (or -15.884716% from expected)
Node(endPoint=/127.0.0.13:9043, hostId=78720bae-4b89-45a8-884a-20f6352b7202, hashCode=14054afa) handled 329439 queries of 1000000 which is 32.943900% of the total query load (or 1.168201% from expected

Run #2
Counts: (16 tokens)
Node(endPoint=/127.0.0.11:9043, hostId=3c57d2e6-f076-4944-8e68-3199252d9132, hashCode=10411ae2) handled 284583 queries of 1000000 which is 28.458300% of the total query load (or 14.625015% from expected)
Node(endPoint=/127.0.0.12:9043, hostId=95f946e2-5007-48e5-953f-c048a812f2f5, hashCode=eeb0871) handled 385667 queries of 1000000 which is 38.566700% of the total query load (or -15.700216% from expected)
Node(endPoint=/127.0.0.13:9043, hostId=129095fb-314a-42a3-8392-d580ca3e5527, hashCode=78959dd) handled 329750 queries of 1000000 which is 32.975000% of the total query load (or 1.074901% from expected)

Run #3
Counts: (16 tokens)
Node(endPoint=/127.0.0.11:9043, hostId=cdde2dfb-259d-4f60-aced-36ba6c07427c, hashCode=994bd47) handled 284326 queries of 1000000 which is 28.432600% of the total query load (or 14.702115% from expected)
Node(endPoint=/127.0.0.12:9043, hostId=a0409f79-5bfb-4dc8-8a20-07c68c6a8ff3, hashCode=2622be1) handled 385996 queries of 1000000 which is 38.599600% of the total query load (or -15.798916% from expected)
Node(endPoint=/127.0.0.13:9043, hostId=36a76255-e02e-4fad-8089-96087f825ef9, hashCode=7717354f) handled 329678 queries of 1000000 which is 32.967800% of the total query load (or 1.096501% from expected)

### 32 tokens ###

Run #1
Counts: (32 tokens)
Node(endPoint=/127.0.0.11:9043, hostId=0f1beb5b-62d2-49cc-8cbd-204dc4346d52, hashCode=29591a4d) handled 374317 queries of 1000000 which is 37.431700% of the total query load (or -12.295212% from expected)
Node(endPoint=/127.0.0.12:9043, hostId=468d97e1-0b4e-43fd-b6b2-d05ebdb15624, hashCode=1e5b0b74) handled 280538 queries of 1000000 which is 28.053800% of the total query load (or 15.838516% from expected)
Node(endPoint=/127.0.0.13:9043, hostId=9d93b462-ae20-4921-9b49-6e493a6f56aa, hashCode=26ee73a3) handled 345145 queries of 1000000 which is 34.514500% of the total query load (or -3.543604% from expected)

Run #2
Counts: (32 tokens)
Node(endPoint=/127.0.0.11:9043, hostId=d74418f3-8c7c-429c-b274-413f30972be1, hashCode=468576e5) handled 373672 queries of 1000000 which is 37.367200% of the total query load (or -12.101712% from expected)
Node(endPoint=/127.0.0.12:9043, hostId=b350efaa-5ceb-4440-ba14-c7ef9f637b90, hashCode=1087ccca) handled 280823 queries of 1000000 which is 28.082300% of the total query load (or 15.753016% from expected)
Node(endPoint=/127.0.0.13:9043, hostId=43b60335-30c6-40d7-954b-314316d8f266, hashCode=20789fe0) handled 345505 queries of 1000000 which is 34.550500% of the total query load (or -3.651604% from expected)

### 64 tokens ###

Run #1
Counts: (64 tokens)
Node(endPoint=/127.0.0.11:9043, hostId=9a89a64d-f7d8-4367-b63a-1aada13a622d, hashCode=50daa05e) handled 387455 queries of 1000000 which is 38.745500% of the total query load (or -16.236616% from expected)
Node(endPoint=/127.0.0.12:9043, hostId=9af169a1-25db-4b9b-9b1e-53e3b0fcf293, hashCode=27b2a005) handled 297719 queries of 1000000 which is 29.771900% of the total query load (or 10.684211% from expected)
Node(endPoint=/127.0.0.13:9043, hostId=e30dcc22-1286-4767-b40d-4524f73e44d6, hashCode=584096e0) handled 314826 queries of 1000000 which is 31.482600% of the total query load (or 5.552106% from expected)

Run #2
Counts: (64 tokens)
Node(endPoint=/127.0.0.11:9043, hostId=1e96ebb6-6ed3-42fd-b1bb-6280c97b4eb7, hashCode=3af1dbdf) handled 386784 queries of 1000000 which is 38.678400% of the total query load (or -16.035316% from expected)
Node(endPoint=/127.0.0.12:9043, hostId=6226b25b-daac-482c-8341-5f7318c6f369, hashCode=6102b27d) handled 298191 queries of 1000000 which is 29.819100% of the total query load (or 10.542611% from expected)
Node(endPoint=/127.0.0.13:9043, hostId=02978d41-a636-4d3e-bb6e-33d0f56fc40a, hashCode=6d7f5eaa) handled 315025 queries of 1000000 which is 31.502500% of the total query load (or 5.492405% from expected)

### 128 tokens ###

Run #1
Counts: (128 tokens)
Node(endPoint=/127.0.0.11:9043, hostId=a3511475-6926-49f1-9a0e-9d7e308efbcb, hashCode=7e49733b) handled 339328 queries of 1000000 which is 33.932800% of the total query load (or -1.798502% from expected)
Node(endPoint=/127.0.0.12:9043, hostId=58baf6ff-d694-4e28-a0a5-006b7e420b70, hashCode=158c50bc) handled 355659 queries of 1000000 which is 35.565900% of the total query load (or -6.697807% from expected)
Node(endPoint=/127.0.0.13:9043, hostId=71703536-61e6-43a8-bc76-496a207313d2, hashCode=5359f56f) handled 305013 queries of 1000000 which is 30.501300% of the total query load (or 8.496008% from expected)

Run #2
Counts: (128 tokens)
Node(endPoint=/127.0.0.11:9043, hostId=bcbee87d-f41c-4d43-90a2-e75bf7774474, hashCode=3b743af3) handled 338867 queries of 1000000 which is 33.886700% of the total query load (or -1.660202% from expected)
Node(endPoint=/127.0.0.12:9043, hostId=9968a400-ebd8-42aa-93f4-f1796442cd90, hashCode=67811a93) handled 355109 queries of 1000000 which is 35.510900% of the total query load (or -6.532807% from expected)
Node(endPoint=/127.0.0.13:9043, hostId=b222b261-cd48-4c5f-8326-c5b9381fba92, hashCode=47ae0cd0) handled 306024 queries of 1000000 which is 30.602400% of the total query load (or 8.192708% from expected)

### 256 tokens ###

Run #1Counts: (256 tokens)
Node(endPoint=/127.0.0.11:9043, hostId=5b2bb6ca-5fc7-4eb8-a1b4-04dd8e06a669, hashCode=36ddad33) handled 337362 queries of 1000000 which is 33.736200% of the total query load (or -1.208701% from expected)
Node(endPoint=/127.0.0.12:9043, hostId=f3a7fae2-2c45-41d9-90dd-4cffe987f938, hashCode=2a99c8d6) handled 345818 queries of 1000000 which is 34.581800% of the total query load (or -3.745504% from expected)
Node(endPoint=/127.0.0.13:9043, hostId=75db2133-06ef-45e8-8d00-55450309f254, hashCode=16f122e8) handled 316820 queries of 1000000 which is 31.682000% of the total query load (or 4.953905% from expected)

Run #2
Counts: (256 tokens)
Node(endPoint=/127.0.0.11:9043, hostId=f05efffa-ebdd-44c4-9d32-f2463e4546db, hashCode=30bade1f) handled 336997 queries of 1000000 which is 33.699700% of the total query load (or -1.099201% from expected)
Node(endPoint=/127.0.0.12:9043, hostId=fc0b92b5-8965-42f3-aa77-dc76e06b0749, hashCode=24b6461b) handled 346448 queries of 1000000 which is 34.644800% of the total query load (or -3.934504% from expected)
Node(endPoint=/127.0.0.13:9043, hostId=55521b7b-031c-457f-82f4-302f4f3d2bda, hashCode=79efea79) handled 316555 queries of 1000000 which is 31.655500% of the total query load (or 5.033405% from expected)

### 512 tokens ###

Run #1
Counts: (512 tokens)
Node(endPoint=/127.0.0.11:9043, hostId=f9345401-bee0-4649-ae90-87d56a4af4ba, hashCode=7d9fce91) handled 338963 queries of 1000000 which is 33.896300% of the total query load (or -1.689002% from expected)
Node(endPoint=/127.0.0.12:9043, hostId=f4d0d396-cf39-45d6-8284-0827f4aeb021, hashCode=4eed6ab4) handled 351231 queries of 1000000 which is 35.123100% of the total query load (or -5.369405% from expected)
Node(endPoint=/127.0.0.13:9043, hostId=061b2228-f049-4c31-a43b-6a0be3afbce7, hashCode=7da07f24) handled 309806 queries of 1000000 which is 30.980600% of the total query load (or 7.058107% from expected)

Run #2
Counts: (512 tokens)
Node(endPoint=/127.0.0.11:9043, hostId=9ddbcbe7-6ff8-4804-9c11-eebdde9029f8, hashCode=7e49733b) handled 340119 queries of 1000000 which is 34.011900% of the total query load (or -2.035802% from expected)
Node(endPoint=/127.0.0.12:9043, hostId=d48ee2b3-f5a5-462f-af19-5b3ac303c858, hashCode=158c50bc) handled 350656 queries of 1000000 which is 35.065600% of the total query load (or -5.196905% from expected)
Node(endPoint=/127.0.0.13:9043, hostId=1183ee78-461d-40e1-bcb5-89887d62e32f, hashCode=5359f56f) handled 309225 queries of 1000000 which is 30.922500% of the total query load (or 7.232407% from expected)
```

Reply:
Thanks @mpenick !

My reading of this is that adding tokens seems to help to some extent with reducing load imbalances, but it's not very efficient. With three nodes even 512 tokens result in up to 7% imbalances.

I suppose it would be valuable to explore the behaviour of the system by:
* run the test with 3,4,5... N nodes
* measure the worst relative query distribution imbalance (relative to expected queries to be handled)
* see if this number decreases with N
* (may be repeat with 8, 64, 512 tokens)

On the other hand, since we're going to recommend round-robin driver settings, this is probably not very critical.